File: angr/procedures/libc/getopt.py
from __future__ import annotations

import logging

import claripy

import angr

l = logging.getLogger(name=__name__)

# struct option { const char *name; int has_arg; int *flag; int val; };
_NO_ARGUMENT = 0
_REQUIRED_ARGUMENT = 1
_OPTIONAL_ARGUMENT = 2

_DASH = ord("-")
_COLON = ord(":")


def _eval_ptr(state, addr):
    """Load a pointer-sized word at ``addr``; return a concrete int or None if symbolic."""
    word = state.memory.load(addr, state.arch.bytes, endness=state.arch.memory_endness)
    if state.solver.symbolic(word):
        return None
    return state.solver.eval(word)


def _concrete_cstr(state, ptr, cap=4096):
    """Read a NUL-terminated string at ``ptr``. Returns bytes (no NUL) or None if any byte is symbolic."""
    out = bytearray()
    for i in range(cap):
        b = state.memory.load(ptr + i, 1)
        if state.solver.symbolic(b):
            return None
        v = state.solver.eval(b)
        if v == 0:
            return bytes(out)
        out.append(v)
    return bytes(out)


def _parse_optstring(s):
    """Return ({char: has_arg}, leading_colon) for an optstring (POSIX/glibc syntax)."""
    opts = {}
    i = 0
    # leading mode chars ('+'/'-' affect ordering, which we do not model)
    while i < len(s) and s[i] in (ord("+"), _DASH):
        i += 1
    leading_colon = i < len(s) and s[i] == _COLON
    if leading_colon:
        i += 1
    while i < len(s):
        c = s[i]
        i += 1
        nargs = _NO_ARGUMENT
        if i < len(s) and s[i] == _COLON:
            nargs = _REQUIRED_ARGUMENT
            i += 1
            if i < len(s) and s[i] == _COLON:
                nargs = _OPTIONAL_ARGUMENT
                i += 1
        opts[c] = nargs
    return opts, leading_colon


class _GetOptBase(angr.SimProcedure):
    """Shared concrete-argv getopt engine.

    Faithful for the common case where ``argv``, the option strings, and the
    relevant bytes are concrete. When anything required is symbolic we fall
    back to an unconstrained return (preserving the pre-SimProc behaviour) so
    we never branch on a symbolic option scan. POSIX non-permuting semantics:
    option scanning stops at the first non-option operand (glibc's '+' /
    POSIXLY_CORRECT mode); argv is never reordered.
    """

    # ---- global-variable plumbing (optind / optarg / optopt / opterr) ----

    def _global_addr(self, name):
        proj = self.project if self.project is not None else self.state.project
        if proj is None:
            return None
        sym = proj.loader.find_symbol(name)
        return sym.rebased_addr if sym is not None else None

    def _store_int(self, name, value):
        addr = self._global_addr(name)
        if addr is not None:
            self.state.memory.store(
                addr,
                claripy.BVV(value, 32),
                endness=self.state.arch.memory_endness,
                inspect=False,
                disable_actions=True,
            )

    def _store_ptr(self, name, value):
        addr = self._global_addr(name)
        if addr is not None:
            self.state.memory.store(
                addr,
                claripy.BVV(value, self.state.arch.bits),
                endness=self.state.arch.memory_endness,
                inspect=False,
                disable_actions=True,
            )

    def _load_cursor(self):
        """Resolve (optind, optchar) honouring a guest reset of ``optind``."""
        plugin_optind = self.state.libc.getopt_optind
        plugin_optchar = self.state.libc.getopt_optchar
        addr = self._global_addr("optind")
        if addr is not None:
            word = self.state.memory.load(addr, 4, endness=self.state.arch.memory_endness)
            if not self.state.solver.symbolic(word):
                guest = self.state.solver.eval(word)
                if guest == 0:
                    return 1, 0  # glibc: optind==0 requests a full reset
                if guest >= 1:
                    # honour a guest rescan (optind reset to a smaller index)
                    return guest, (plugin_optchar if guest == plugin_optind else 0)
        return plugin_optind, plugin_optchar

    def _save_cursor(self, optind, optchar):
        self.state.libc.getopt_optind = optind
        self.state.libc.getopt_optchar = optchar
        self._store_int("optind", optind)

    @staticmethod
    def _unconstrained():
        return claripy.BVS("getopt", 32)

    @staticmethod
    def _ret_char(c):
        return claripy.BVV(c, 32)

    @staticmethod
    def _ret_done():
        return claripy.BVV(0xFFFFFFFF, 32)  # -1

    # ---- long-option support ----

    def _handle_long(
        self,
        name_part,
        prefix_len,
        longopts_addr,
        has_longindex,
        longindex_addr,
        optind,
        argc,
        argv_ptr,
        leading_colon,
    ):
        """Match a long option (name, possibly ``name=value``). Returns a BV
        result, or None when there is no match (so getopt_long_only can fall
        back to short-option parsing)."""
        state = self.state
        ps = state.arch.bytes
        stride = 4 * ps  # name(ps) has_arg(int,aligned) flag(ptr) val(int) -> 4*ps with padding

        eq = name_part.find(b"=")
        if eq >= 0:
            name = name_part[:eq]
            inline_val = name_part[eq + 1 :]
        else:
            name = name_part
            inline_val = None

        # scan the option table for an exact match, then a unique prefix match
        ptr = longopts_addr
        exact = None
        prefixes = []
        idx = 0
        while True:
            name_ptr = _eval_ptr(state, ptr)
            if name_ptr is None:
                return self._unconstrained()
            if name_ptr == 0:
                break
            optname = _concrete_cstr(state, name_ptr)
            if optname is None:
                return self._unconstrained()
            if optname == name:
                exact = (idx, ptr)
                break
            if name and optname.startswith(name):
                prefixes.append((idx, ptr))
            idx += 1
            ptr += stride

        if exact is not None:
            match = exact
        elif len(prefixes) == 1:
            match = prefixes[0]
        elif len(prefixes) > 1:
            # ambiguous abbreviation
            self._store_int("optopt", 0)
            self._store_ptr("optarg", 0)
            self._save_cursor(optind + 1, 0)
            return self._ret_char(ord("?"))
        else:
            return None  # no match

        match_idx, match_ptr = match
        has_arg = state.memory.load(match_ptr + ps, 4, endness=state.arch.memory_endness)
        has_arg = state.solver.eval(has_arg) if not state.solver.symbolic(has_arg) else _NO_ARGUMENT
        flag_ptr = _eval_ptr(state, match_ptr + 2 * ps)
        val_word = state.memory.load(match_ptr + 3 * ps, 4, endness=state.arch.memory_endness)
        val = state.solver.eval(val_word) if not state.solver.symbolic(val_word) else 0

        if has_longindex and longindex_addr:
            state.memory.store(
                longindex_addr,
                claripy.BVV(match_idx, 32),
                endness=state.arch.memory_endness,
                inspect=False,
                disable_actions=True,
            )

        # resolve the option-argument
        new_optind = optind + 1
        optarg = 0
        if has_arg == _REQUIRED_ARGUMENT:
            if inline_val is not None:
                # optarg points just past the '=' in the original argv element
                elem_ptr = _eval_ptr(state, argv_ptr + optind * ps)
                optarg = (elem_ptr + prefix_len + (len(name_part) - len(inline_val))) if elem_ptr is not None else 0
            elif optind + 1 < argc:
                optarg = _eval_ptr(state, argv_ptr + (optind + 1) * ps) or 0
                new_optind = optind + 2
            else:
                # missing required argument
                self._store_int("optopt", val if flag_ptr == 0 else 0)
                self._store_ptr("optarg", 0)
                self._save_cursor(new_optind, 0)
                return self._ret_char(_COLON if leading_colon else ord("?"))
        elif has_arg == _OPTIONAL_ARGUMENT and inline_val is not None:
            elem_ptr = _eval_ptr(state, argv_ptr + optind * ps)
            optarg = (elem_ptr + (len(name_part) - len(inline_val))) if elem_ptr is not None else 0

        self._store_ptr("optarg", optarg)
        self._save_cursor(new_optind, 0)

        if flag_ptr:
            state.memory.store(
                flag_ptr,
                claripy.BVV(val, 32),
                endness=state.arch.memory_endness,
                inspect=False,
                disable_actions=True,
            )
            return self._ret_char(0)
        return self._ret_char(val)

    # ---- main engine ----

    def _getopt(self, argc, argv, optstring, longopts=None, longindex=None, long_only=False):
        state = self.state
        ps = state.arch.bytes

        if state.solver.symbolic(argc):
            return self._unconstrained()
        argc = state.solver.eval(argc)
        if state.solver.symbolic(argv):
            return self._unconstrained()
        argv_ptr = state.solver.eval(argv)
        if state.solver.symbolic(optstring):
            return self._unconstrained()
        optstring_bytes = _concrete_cstr(state, state.solver.eval(optstring))
        if optstring_bytes is None:
            return self._unconstrained()

        longopts_addr = None
        if longopts is not None and not state.solver.symbolic(longopts):
            longopts_addr = state.solver.eval(longopts)
            if longopts_addr == 0:
                longopts_addr = None
        longindex_addr = None
        has_longindex = longindex is not None and not state.solver.symbolic(longindex)
        if has_longindex:
            longindex_addr = state.solver.eval(longindex)
            has_longindex = longindex_addr != 0

        opts, leading_colon = _parse_optstring(optstring_bytes)
        optind, optchar = self._load_cursor()

        if optind >= argc:
            self._save_cursor(optind, 0)
            self._store_ptr("optarg", 0)
            return self._ret_done()

        elem_ptr = _eval_ptr(state, argv_ptr + optind * ps)
        if elem_ptr is None:
            return self._unconstrained()
        if elem_ptr == 0:
            self._save_cursor(optind, 0)
            return self._ret_done()
        arg = _concrete_cstr(state, elem_ptr)
        if arg is None:
            return self._unconstrained()

        if optchar == 0:
            if not arg or arg[0] != _DASH or arg == b"-":
                # non-option operand -> stop (non-permuting)
                self._save_cursor(optind, 0)
                return self._ret_done()
            if arg == b"--":
                self._save_cursor(optind + 1, 0)
                return self._ret_done()
            if longopts_addr is not None:
                long_attempt = None
                prefix_len = 0
                if arg.startswith(b"--"):
                    long_attempt = arg[2:]
                    prefix_len = 2
                elif long_only:
                    long_attempt = arg[1:]
                    prefix_len = 1
                if long_attempt is not None:
                    res = self._handle_long(
                        long_attempt,
                        prefix_len,
                        longopts_addr,
                        has_longindex,
                        longindex_addr,
                        optind,
                        argc,
                        argv_ptr,
                        leading_colon,
                    )
                    if res is not None:
                        return res
                    if arg.startswith(b"--"):
                        # unknown "--long" never falls back to short parsing
                        self._store_int("optopt", 0)
                        self._store_ptr("optarg", 0)
                        self._save_cursor(optind + 1, 0)
                        return self._ret_char(ord("?"))
                    # long_only single-dash with no match -> try short below
            optchar = 1

        c = arg[optchar]
        spec = opts.get(c)
        if spec is None or c == _COLON:
            self._store_int("optopt", c)
            self._store_ptr("optarg", 0)
            optchar += 1
            if optchar >= len(arg):
                optind += 1
                optchar = 0
            self._save_cursor(optind, optchar)
            return self._ret_char(ord("?"))

        if spec == _NO_ARGUMENT:
            self._store_ptr("optarg", 0)
            optchar += 1
            if optchar >= len(arg):
                optind += 1
                optchar = 0
            self._save_cursor(optind, optchar)
            return self._ret_char(c)

        # option takes an argument (required or optional)
        if optchar + 1 < len(arg):
            self._store_ptr("optarg", elem_ptr + optchar + 1)
            self._save_cursor(optind + 1, 0)
            return self._ret_char(c)
        if spec == _OPTIONAL_ARGUMENT:
            self._store_ptr("optarg", 0)
            self._save_cursor(optind + 1, 0)
            return self._ret_char(c)
        # required argument from the next argv element
        if optind + 1 < argc:
            next_ptr = _eval_ptr(state, argv_ptr + (optind + 1) * ps)
            if next_ptr is None:
                return self._unconstrained()
            self._store_ptr("optarg", next_ptr)
            self._save_cursor(optind + 2, 0)
            return self._ret_char(c)
        # missing required argument
        self._store_int("optopt", c)
        self._store_ptr("optarg", 0)
        self._save_cursor(optind + 1, 0)
        return self._ret_char(_COLON if leading_colon else ord("?"))


class getopt(_GetOptBase):
    # pylint:disable=arguments-differ
    def run(self, argc, argv, optstring):
        return self._getopt(argc, argv, optstring)


class getopt_long(_GetOptBase):
    # pylint:disable=arguments-differ
    def run(self, argc, argv, optstring, longopts, longindex):
        return self._getopt(argc, argv, optstring, longopts=longopts, longindex=longindex)


class getopt_long_only(_GetOptBase):
    # pylint:disable=arguments-differ
    def run(self, argc, argv, optstring, longopts, longindex):
        return self._getopt(argc, argv, optstring, longopts=longopts, longindex=longindex, long_only=True)

File: native/angr/src/procedures/getopt.rs
//! Native `getopt(3)` short-option engine.
//!
//! Faithful Rust port of the short-option path of
//! `angr/procedures/libc/getopt.py::_GetOptBase._getopt`. Registered only for
//! the 3-argument `getopt` (no long-option variants): `getopt_long` /
//! `getopt_long_only` stay on the Python proc, which owns the option-table
//! walk and ambiguous-abbreviation handling.
//!
//! Contract — **match Python byte-for-byte, or defer** (return
//! `ProcedureError` → Python fallback). We defer whenever any required input
//! is symbolic (argc, argv, an argv element pointer, the optstring, or any
//! scanned byte), mirroring Python's `_unconstrained()` branch which produces a
//! fresh symbolic return rather than committing to a concrete scan. Deferring
//! lets the Python proc produce that unconstrained value itself.
//!
//! Cross-call cursor state (`optind` / `optchar`) lives on `RustSimState`
//! (bhk0a.3 cursor fields). The loader-resolved guest addresses of the
//! `optind` / `optarg` / `optopt` extern globals are pushed from Python into
//! `RustSimState::getopt_extern` (bhk0a.1); we write the updated values back
//! there so the guest program reads them like real getopt would. A `None`
//! address (symbol absent) just skips that store — exactly as Python's
//! `_store_int` / `_store_ptr` skip when `_global_addr` returns `None`.
//!
//! **Panic policy / enforcement (angr-qwyti.11, angr-9ke6b.212):** `argv`, the
//! optstring and the saved cursor are all guest-controlled, so this module
//! carries `#![deny(clippy::unwrap_used, clippy::expect_used)]` and every
//! unresolvable value returns a `ProcedureError` that falls back to the Python
//! proc. There are no `unwrap`/`expect` sites left: the optstring lookup binds
//! its hit in the match arm that proves it (see the `spec` match), and the
//! cursor read already used a checked `arg.get(..)` (angr-qwyti.19).
#![deny(clippy::unwrap_used, clippy::expect_used)]

use super::strings::{MAX_STRING_SCAN, scan_concrete_until_null};
use super::{ProcedureError, arch_word};
use crate::state::RustSimState;
use crate::symbolic::RustBV;

const NO_ARGUMENT: u8 = 0;
const REQUIRED_ARGUMENT: u8 = 1;
const OPTIONAL_ARGUMENT: u8 = 2;

const DASH: u8 = b'-';
const COLON: u8 = b':';

/// Parse an optstring into `(char -> has_arg, leading_colon)` per POSIX/glibc
/// syntax. Leading `+`/`-` mode chars (ordering, which we do not model) are
/// skipped; a leading `:` selects the missing-arg `':'` return.
fn parse_optstring(s: &[u8]) -> (std::collections::HashMap<u8, u8>, bool) {
    let mut opts = std::collections::HashMap::new();
    let mut i = 0;
    while i < s.len() && (s[i] == b'+' || s[i] == DASH) {
        i += 1;
    }
    let leading_colon = i < s.len() && s[i] == COLON;
    if leading_colon {
        i += 1;
    }
    while i < s.len() {
        let c = s[i];
        i += 1;
        let mut nargs = NO_ARGUMENT;
        if i < s.len() && s[i] == COLON {
            nargs = REQUIRED_ARGUMENT;
            i += 1;
            if i < s.len() && s[i] == COLON {
                nargs = OPTIONAL_ARGUMENT;
                i += 1;
            }
        }
        opts.insert(c, nargs);
    }
    (opts, leading_colon)
}

/// Load a pointer-sized word; `Ok(None)` if symbolic (caller defers).
fn eval_ptr(state: &RustSimState, addr: u64) -> Result<Option<u64>, ProcedureError> {
    let ps = state.arch().bytes();
    Ok(state.memory_load(addr, ps)?.as_u64())
}

/// Store a 32-bit int at the extern global `addr` if present (else no-op),
/// mirroring Python's `_store_int`.
fn store_int(
    state: &mut RustSimState,
    addr: Option<u64>,
    value: u32,
) -> Result<(), ProcedureError> {
    if let Some(a) = addr {
        state.memory_store(a, RustBV::concrete(value as u128, 32))?;
    }
    Ok(())
}

/// Store a pointer-sized word at the extern global `addr` if present,
/// mirroring Python's `_store_ptr`.
fn store_ptr(
    state: &mut RustSimState,
    addr: Option<u64>,
    value: u64,
) -> Result<(), ProcedureError> {
    if let Some(a) = addr {
        state.memory_store(a, arch_word(state, value))?;
    }
    Ok(())
}

/// Resolve `(optind, optchar)`, honouring a guest reset of `optind` exactly as
/// Python's `_load_cursor`.
fn load_cursor(
    state: &RustSimState,
    optind_addr: Option<u64>,
) -> Result<(u32, u32), ProcedureError> {
    let (plugin_optind, plugin_optchar) = state.getopt_cursor();
    if let Some(addr) = optind_addr {
        let word = state.memory_load(addr, 4)?;
        if let Some(guest) = word.as_u64() {
            let guest = guest as u32;
            if guest == 0 {
                return Ok((1, 0)); // glibc: optind==0 requests a full reset
            }
            // guest >= 1: honour a guest rescan (optind reset to a smaller index)
            let oc = if guest == plugin_optind {
                plugin_optchar
            } else {
                0
            };
            return Ok((guest, oc));
        }
    }
    Ok((plugin_optind, plugin_optchar))
}

/// Persist `(optind, optchar)` to the cursor fields and write `optind` back to
/// guest memory, mirroring Python's `_save_cursor`.
fn save_cursor(
    state: &mut RustSimState,
    optind_addr: Option<u64>,
    optind: u32,
    optchar: u32,
) -> Result<(), ProcedureError> {
    state.set_getopt_cursor(optind, optchar);
    store_int(state, optind_addr, optind)
}

/// `int getopt(int argc, char *const argv[], const char *optstring)`.
#[inline]
fn ret(value: u32) -> Result<Option<RustBV>, ProcedureError> {
    Ok(Some(RustBV::concrete(value as u128, 32)))
}

crate::declare_proc! {
    /// getopt: concrete-argv short-option engine. Defers to Python for any
    /// symbolic input (matching the Python proc's unconstrained branch).
    ///
    /// ```c
    /// int getopt(int argc, char *const argv[], const char *optstring);
    /// ```
    name = "getopt",
    struct = NativeGetopt,
    args = [argc: concrete, argv_ptr: concrete, optstring_ptr: concrete],
    call |state| {
        let ps = state.arch().bytes() as u64;
        let extern_addrs = state.getopt_extern();
        let optind_addr = extern_addrs.optind;
        let optarg_addr = extern_addrs.optarg;
        let optopt_addr = extern_addrs.optopt;

        let argc = argc as u32;

        let optstring = scan_concrete_until_null(state, optstring_ptr, MAX_STRING_SCAN, "optstring")?;
        let (opts, leading_colon) = parse_optstring(&optstring);

        let (mut optind, mut optchar) = load_cursor(state, optind_addr)?;

        if optind >= argc {
            save_cursor(state, optind_addr, optind, 0)?;
            store_ptr(state, optarg_addr, 0)?;
            return ret(0xFFFF_FFFF); // -1
        }

        // `argv_ptr` is a guest-controlled address, so every address computation
        // below uses `wrapping_add` per the convention documented in memcpy.rs:
        // `[profile.release]` disables overflow checks, so a plain `+` panics
        // under the release-checked test profile and silently wraps in the
        // shipped `.so` — a wrong address rather than a loud failure.
        // (`optind`/`optchar` are u32, so their `as u64` scaling cannot itself
        // overflow: u32::MAX * 8 fits in u64.)
        let elem_ptr = match eval_ptr(state, argv_ptr.wrapping_add(optind as u64 * ps))? {
            Some(p) => p,
            None => return Err(ProcedureError::SymbolicArgument("argv element".into())),
        };
        if elem_ptr == 0 {
            save_cursor(state, optind_addr, optind, 0)?;
            return ret(0xFFFF_FFFF);
        }
        let arg = scan_concrete_until_null(state, elem_ptr, MAX_STRING_SCAN, "argv string")?;

        if optchar == 0 {
            // non-option operand "-" or a non-dash element -> stop (non-permuting)
            if arg.is_empty() || arg[0] != DASH || arg == [DASH] {
                save_cursor(state, optind_addr, optind, 0)?;
                return ret(0xFFFF_FFFF);
            }
            if arg == [DASH, DASH] {
                save_cursor(state, optind_addr, optind + 1, 0)?;
                return ret(0xFFFF_FFFF);
            }
            // long-option handling lives in the Python proc; the native proc is
            // registered only for plain `getopt`, so longopts is always absent.
            optchar = 1;
        }

        // Defense-in-depth (angr-qwyti.19): `optchar` is engine cursor state
        // restored from a prior getopt call. If the guest mutated this argv
        // element in its own memory to be shorter since then, the saved cursor
        // can exceed the freshly-scanned `arg` length. Use a checked accessor
        // and defer to the Python getopt proc rather than panic-indexing.
        let Some(&c) = arg.get(optchar as usize) else {
            return Err(ProcedureError::Other(
                "getopt cursor past end of argv element (shrunk argv)".to_string(),
            ));
        };
        // `c` is a guest byte, so the optstring lookup genuinely can miss —
        // binding the hit in the match arm keeps the unknown-option path and
        // the recognized-option path in one place, with no second lookup and no
        // `unwrap` on what the miss branch just ruled out.
        let spec = match opts.get(&c).copied() {
            Some(spec) if c != COLON => spec,
            // Unknown option (or a bare `:`, which is never an option itself):
            // report it through `optopt` and return '?'.
            _ => {
                store_int(state, optopt_addr, c as u32)?;
                store_ptr(state, optarg_addr, 0)?;
                optchar += 1;
                if optchar as usize >= arg.len() {
                    optind += 1;
                    optchar = 0;
                }
                save_cursor(state, optind_addr, optind, optchar)?;
                return ret(b'?' as u32);
            }
        };

        if spec == NO_ARGUMENT {
            store_ptr(state, optarg_addr, 0)?;
            optchar += 1;
            if optchar as usize >= arg.len() {
                optind += 1;
                optchar = 0;
            }
            save_cursor(state, optind_addr, optind, optchar)?;
            return ret(c as u32);
        }

        // option takes an argument (required or optional)
        if (optchar as usize) + 1 < arg.len() {
            store_ptr(state, optarg_addr, elem_ptr.wrapping_add(optchar as u64 + 1))?;
            save_cursor(state, optind_addr, optind + 1, 0)?;
            return ret(c as u32);
        }
        if spec == OPTIONAL_ARGUMENT {
            store_ptr(state, optarg_addr, 0)?;
            save_cursor(state, optind_addr, optind + 1, 0)?;
            return ret(c as u32);
        }
        // required argument from the next argv element
        if optind + 1 < argc {
            let next_ptr = match eval_ptr(state, argv_ptr.wrapping_add((optind as u64 + 1) * ps))? {
                Some(p) => p,
                None => return Err(ProcedureError::SymbolicArgument("argv next".into())),
            };
            store_ptr(state, optarg_addr, next_ptr)?;
            save_cursor(state, optind_addr, optind + 2, 0)?;
            return ret(c as u32);
        }
        // missing required argument
        store_int(state, optopt_addr, c as u32)?;
        store_ptr(state, optarg_addr, 0)?;
        save_cursor(state, optind_addr, optind + 1, 0)?;
        ret(if leading_colon { COLON as u32 } else { b'?' as u32 })
    }
}

test_submod!("getopt_tests.rs" => getopt_tests);
